### PR TITLE
#68 Schedule tab implemented

### PR DESCRIPTION
**Pull request by YuraRiabov**
Schedule and schedule item entities created, respective components on the frontend added, creating new availability from frontend implemented, creating and updating availability updated to allow schedule functionality
Screenshot_536

### CODE REVIEW

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 1]
```diff
@@ -17,14 +18,19 @@ public class AvailabilityService : BaseService, IAvailabilityService
...
+            var availabilitySlots = (await _context.AvailabilitySlots
                 .Include(x => x.Members)
                     .ThenInclude(x => x.User)
                 .Include(x => x.Author)
                 .Include(x => x.Team)
-                .Where(x => x.CreatedBy == id && x.Members.Any(x => x.UserId == id))
+                .Include(x => x.Schedule)
+                    .ThenInclude(s => s.ScheduleItems)
+                .Where(x => x.CreatedBy == id || x.Members.Any(x => x.UserId == id))
+                .ToListAsync())
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: Why did you remove the projection?

Why do we need Schedule here?

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 2]
```diff
@@ -65,17 +75,35 @@ public async Task CreateAvailabilitySlot(NewAvailabilitySlotDto slotDto)
             {
                 var advancedSettings = _mapper.Map<AdvancedSlotSettings>(slotDto.AdvancedSettings);
                 advancedSettings.AvailabilitySlot = entity;
+                advancedSettings.StartDate = advancedSettings.StartDate == DateTimeOffset.MinValue
+                    ? DateTimeOffset.Now
+                    : advancedSettings.StartDate;
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: Please use .AfterMap

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 3]
```diff
@@ -65,17 +75,35 @@ public async Task CreateAvailabilitySlot(NewAvailabilitySlotDto slotDto)
...
                 await _context.AdvancedSlotSettings.AddAsync(advancedSettings);
                 entity.AdvancedSlotSettings = advancedSettings;
             }
 
+            var schedule = _mapper.Map<Schedule>(slotDto.Schedule);
+            schedule.AvailabilitySlot = entity;
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: See the previous coment

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 4]
```diff
@@ -65,17 +75,35 @@ public async Task CreateAvailabilitySlot(NewAvailabilitySlotDto slotDto)
...
+            _context.Schedules.Add(schedule);
+            entity.Schedule = schedule;
+            
+            foreach (var itemDto in slotDto.Schedule.ScheduleItems)
+            {
+                var item = _mapper.Map<ScheduleItem>(itemDto);
+                item.Schedule = schedule;
+                _context.ScheduleItems.Add(item);
+                schedule.ScheduleItems.Add(item);
+            }
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: Move this to a separate method

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs [hunk 5]
```diff
@@ -112,6 +142,12 @@ public async Task<AvailabilitySlotDto> UpdateAvailabilitySlot(long id, UpdateAva
             {
                 _context.Remove(availabilitySlot.AdvancedSlotSettings);
             }
+
+            _mapper.Map(updateAvailabilityDto.Schedule, availabilitySlot.Schedule);
+            for (int i = 0; i < updateAvailabilityDto.Schedule.ScheduleItems.Count; i++)
+            {
+                _mapper.Map(updateAvailabilityDto.Schedule.ScheduleItems[i], availabilitySlot.Schedule.ScheduleItems.ToList()[i]);
+            }
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: Please refactor these

File: frontend/src/app/core/enums/location-type.enum.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export enum LocationType {
+    Zoom,
+    GoogleMeet,
+    Office
+}
```
Comment (tatianahutii) [inline frontend/src/app/core/enums/location-type.enum.ts]: I believe we already have this enum

File: frontend/src/app/core/models/new-availability-slot/INewEventDetails.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+export interface INewEventDetails {
+    link: string;
+    welcomeMessage: string;
+    language: string;
+    bookingsPerDay: number;
+    allowToAddGuests: boolean;
+    passwordProtectionIsUsed: boolean;
+    passwordProtection: string | null;
```
Comment (tatianahutii) [inline frontend/src/app/core/models/new-availability-slot/INewEventDetails.ts]:
```suggestion
    passwordProtection?: string;
```

File: frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts [hunk 8]
```diff
@@ -1,16 +1,91 @@
...
     styleUrls: ['./new-availability-page.component.sass'],
 })
 export class NewAvailabilityPageComponent {
-    // eslint-disable-next-line no-empty-function
-    constructor(private router: Router) { }
+    constructor(
+        private router: Router,
+        private slotService: AvailabilitySlotService,
+        private notifications: NotificationService,
+        // eslint-disable-next-line no-empty-function
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts]: please delete this and use BaseComponent

File: frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts [hunk 9]
```diff
@@ -1,16 +1,91 @@
...
+    ) { }
+
+    public currentUserId: number = 2;
+
+    private unsubscribe$ = new Subject<void>();
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts]: Please delete this

File: frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts [hunk 10]
```diff
@@ -1,16 +1,91 @@
...
+            },
+            eventDetails: {
+                link: eventDetails.linkChoice,
+                welcomeMessage: eventDetails.welcomeMessage,
+                language: eventDetails.languageSelect,
+                bookingsPerDay: eventDetails.allowBookingSelect,
+                allowToAddGuests: eventDetails.isAllowBooker,
+                passwordProtectionIsUsed: eventDetails.passwordProtect,
+                passwordProtection: eventDetails.passwordInput,
+                timeZoneVisibility: eventDetails.zoneChoice,
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts]: Why do you need this mapping?

File: frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts [hunk 11]
```diff
@@ -1,16 +1,91 @@
...
+        const newAvailability: INewAvailability = {
+            generalDetails: {
+                locationType: general.meetingLocation,
+                name: general.meetingName,
+                type: SlotType.Personal,
+                size: general.slotSize,
+                frequency: general.slotFrequency,
+                isEnabled: this.newAvailabilityComponent.isActive,
+                isVisible: general.hideFromCommon,
+            },
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts]:
```suggestion
            generalDetails: {
                 ...general,
                type: SlotType.Personal,
                isEnabled: this.newAvailabilityComponent.isActive,
            },
```

File: frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts [hunk 12]
```diff
@@ -1,16 +1,91 @@
...
+                welcomeMessage: eventDetails.welcomeMessage,
+                language: eventDetails.languageSelect,
+                bookingsPerDay: eventDetails.allowBookingSelect,
+                allowToAddGuests: eventDetails.isAllowBooker,
+                passwordProtectionIsUsed: eventDetails.passwordProtect,
+                passwordProtection: eventDetails.passwordInput,
+                timeZoneVisibility: eventDetails.zoneChoice,
+            },
+            advancedSettings,
+            createdBy: this.currentUserId,
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts]: We need to set this is the backend

File: frontend/src/app/modules/availability/new-slot/schedule-list-item/schedule-list-item.component.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    public startValue: string;
+
+    public endValue: string;
+
+    public onDateChange($event: any, isStart: boolean) {
+        if (isStart) {
+            this.item.start = `${$event.target.value}:00`;
+        } else {
+            this.item.end = `${$event.target.value}:00`;
+        }
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule-list-item/schedule-list-item.component.ts]: you can create string value before if/else

File: frontend/src/app/modules/availability/new-slot/schedule-list-item/schedule-list-item.component.ts [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Output() public itemChange: EventEmitter<IScheduleItem> = new EventEmitter<IScheduleItem>();
+
+    @Input() public displayDay: string;
+
+    public startValue: string;
+
+    public endValue: string;
+
+    public onDateChange($event: any, isStart: boolean) {
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule-list-item/schedule-list-item.component.ts]: Please do not use any

File: frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+            withTeamMembers: false,
+            scheduleItems: getScheduleItems(),
+        };
+    }
+
+    public slot?: IAvailabilitySlot;
+
+    public schedule: ISchedule;
+
+    public displayDays: string[] = ['Sun', 'Mon', 'Tue', 'Wen', 'Thu', 'Fri', 'Sat'];
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts]: Please create enum or helper for this

File: frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+    public timeZones: Map<number, string> = getPossibleTimeZones();
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts]:
```suggestion
    readonly timeZones: Map<number, string> = getPossibleTimeZones();
```

File: frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    ngOnInit(): void {
+        this.selectedTimeZone = this.timeZones.get(this.schedule.timeZone) as string;
+    }
+
+    public changeTimeZone() {
+        this.schedule.timeZone = this.getSelectedTimeZone();
+    }
+
+    private getSelectedTimeZone() {
+        return [...this.timeZones].find(([, val]) => val === this.selectedTimeZone)![0];
```
Comment (tatianahutii) [inline frontend/src/app/modules/availability/new-slot/schedule/schedule.component.ts]: Please refactor this

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 1 above)
Comment (YuraRiabov) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: We need schedule here to be able to edit it using data from this dto

Comment (tatianahutii) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: I don't think so
This method is for loading all slots, not for some specific

File: frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts (hunk 10 above)
Comment (YuraRiabov) [reply frontend/src/app/modules/availability/new-slot/new-availability-page/new-availability-page.component.ts]: Frontend interfaces don't match backend dtos in terms of property names and changing them would require changing update availability dtos and their mapping as well

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 1 above)
Comment (YuraRiabov) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: Still it loads all the slot's data except for schedule. We might need schedule not only for editing in the future, for example displaying slot for external booking

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs (hunk 4 above)
Comment (YuraRiabov) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/AvailabilityService.cs]: We don't even need this btw

**Review by tatianahutii — APPROVED** [review]
LGTM
